database/migrations/community/2024_12_16_000003_update_forumtopiccomment_table.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+return new class() extends Migration {
+    public function up(): void
+    {
+        Schema::table('ForumTopicComment', function (Blueprint $table) {
+            $table->dropForeign(['author_id']);
+        });
+
+        Schema::rename('ForumTopicComment', 'forum_topic_comments');
+
+        Schema::table('forum_topic_comments', function (Blueprint $table) {
+            $table->renameColumn('ID', 'id');
+            $table->renameColumn('ForumTopicID', 'forum_topic_id');
+            $table->renameColumn('Payload', 'body');
+            $table->renameColumn('DateCreated', 'created_at');
+            $table->renameColumn('DateModified', 'updated_at');
+            $table->renameColumn('Authorised', 'is_authorized');
+        });
+
+        Schema::table('forum_topic_comments', function (Blueprint $table) {
+            $table->foreign('forum_topic_id')->references('id')->on('forum_topics')->onDelete('set null');
```

Comment:
This is failing too:
```
  SQLSTATE[23000]: Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails (`retroachievements-web`.`#sql-alter-1-465`, CONSTRAINT `forum_topic_comments_forum_topic_id_foreign` FOREIGN KEY (`forum_topic_id`) REFERENCES `forum_topics` (`id`) ON DELETE SET NULL) (Connection: mysql, SQL: alter table `forum_topic_comments` add constraint `forum_topic_comments_forum_topic_id_foreign` foreign key (`forum_topic_id`) references `forum_topics` (`id`) on delete set null)
```

Assuming I did my query right, there's a lot of orphaned comments.
```
mysql> select count(distinct ftc.forum_topic_id) from forum_topic_comments ftc left join forum_topics ft on ft.id=ftc.forum_topic_id where ft.id is null;
+------------------------------------+
| count(distinct ftc.forum_topic_id) |
+------------------------------------+
|                               2358 |
+------------------------------------+
```

Reply:
I normally don't like to mutate data in migrations, but I'm beginning to lead towards deletion of data, especially orphaned data, probably being appropriate. In latest I've added:
```php
Schema::table('forum_topic_comments', function (Blueprint $table) {
    // Delete all orphaned comments before adding a foreign key.
    DB::table('forum_topic_comments')
        ->whereNotExists(function ($query) {
            $query->select(DB::raw(1))
                ->from('forum_topics')
                ->whereColumn('forum_topics.id', 'forum_topic_comments.forum_topic_id');
        })
        ->delete();
});
```
to `up()`. There's nothing to change in `down()`, as the data isn't accessible anyway.

Reply:
In the past, we've just done the sanitation with raw DB queries.

Reply:
That's fine too -- this should be the query we need to run pre-deploy:
```sql
DELETE FROM ForumTopicComment 
WHERE NOT EXISTS (
    SELECT 1 
    FROM ForumTopic 
    WHERE ForumTopic.ID = ForumTopicComment.ForumTopicID
);
```

I'll put this in the PR description.